src/stream/src/executor/merge.rs

```diff
@@ -276,8 +375,12 @@ impl Stream for SelectReceivers {
                         self.last_base = (idx + 1) % self.upstreams.len();
                         return Poll::Ready(Some(Ok(message)));
                     }
-                    Some(Ok(Message::Watermark(_))) => {
-                        todo!("https://github.com/risingwavelabs/risingwave/issues/6042")
+                    Some(Ok(Message::Watermark(watermark))) => {
+                        if let Some(watermark) = self.handle_watermark(actor_id, watermark) {
+                            self.last_base = idx;
```

Comment:
> Why do we keep the current index here? 🤔

to make behavior of if..then branch consistent with else branch

Reply:
Should it be `(idx + 1) % self.upstreams.len()`? 🤔

Reply:
> Should it be `(idx + 1) % self.upstreams.len()`? 🤔

We can use `(idx + 1) % self.upstreams.len()` certainly if you insist. However, the else branch will do the same thing as `self.last_base = idx;` and it can not be changed.

Reply:
Did't quite get that 🥵. Receiving and yielding a watermark message is like yielding a chunk, and if we can't emit the watermark, it's like we're `Pending`. So shouldn't we always increase the `poll_count` (or `idx +1`) in both cases? Note that if we can't get `Some` after handling, then it must be blocked by other upstreams. Polling other upstreams also make more sense.

Reply:
> Did't quite get that 🥵. Receiving and yielding a watermark message is like yielding a chunk, and if we can't emit the watermark, it's like we're `Pending`. So shouldn't we always increase the `poll_count` (or `idx +1`) in both cases? Note that if we can't get `Some` after handling, then it must be blocked by other upstreams. Polling other upstreams also make more sense.

We can't increase the `poll_count` because there might be a Chunk just after the Watermark.

Reply:
> Did't quite get that 🥵. Receiving and yielding a watermark message is like yielding a chunk, and if we can't emit the watermark, it's like we're `Pending`. So shouldn't we always increase the `poll_count` (or `idx +1`) in both cases? Note that if we can't get `Some` after handling, then it must be blocked by other upstreams. Polling other upstreams also make more sense.

We can increase the `poll_count` *only if* this upstream is `Poll::Pending`

Reply:
> because there might be a Chunk just after the Watermark.

Then I believe there also "might be" another chunk after a **chunk**. 🤣 So why don't we keep the index for the next polling?

Reply:
> We can increase the `poll_count` _only if_ this upstream is `Poll::Pending`

The polling order really has little impact on the performance, as long as we can make it roughly balanced. So I believe there's no "only if" for this case, and both actually LGTM.

Reply:
> > because there might be a Chunk just after the Watermark.
> 
> Then I believe there also "might be" another chunk after a **chunk**. 🤣 So why don't we keep the index for the next polling?

It's different from the case that chunk after a `chunk`. We can fetch the chunk after a `chunk` in next polling, without returning *any* `Pending`. However, if we do not increase `poll_count` in else branch, we will return a `Pending`. It is not permitted.

Reply:
> > We can increase the `poll_count` _only if_ this upstream is `Poll::Pending`
> 
> The polling order really has little impact on the performance, as long as we can make it roughly balanced. So I believe there's no "only if" for this case, and both actually LGTM.

ok I have just altered the strategy in then branch. However I cannot alter in else branch.

Reply:
> We can't increase the poll_count because there might be a Chunk just after the Watermark.

In the `WatermarkFilter` implementation, it will send a `Chunk` and then a watermark immediately after the chunk.

Reply:
> However I cannot alter in else branch.

Agree. Registering more wakers by polling a `Pending` from upstreams will always be better.